localstack/services/awslambda/invocation/lambda_models.py

```diff
@@ -71,25 +75,81 @@ class Invocation:
 
 @dataclasses.dataclass(frozen=True)
 class S3Code:
+    id: str
     s3_bucket: str
     s3_key: str
     s3_object_version: str | None
     code_sha256: str
     code_size: int
+    _disk_lock: threading.RLock = dataclasses.field(default_factory=threading.RLock)
 
-    def get_lambda_archive(self) -> bytes:
+    def _download_archive_to_file(self, target_file: IO) -> None:
+        """
+        Download the code archive into a given file
+
+        :param target_file: File the code archive should be downloaded into (IO object)
+        """
         s3_client: "S3Client" = aws_stack.connect_to_service("s3", region_name="us-east-1")
-        kwargs = {"VersionId": self.s3_object_version} if self.s3_object_version else {}
-        return s3_client.get_object(Bucket=self.s3_bucket, Key=self.s3_key, **kwargs)["Body"].read()
+        extra_args = {"VersionId": self.s3_object_version} if self.s3_object_version else {}
+        s3_client.download_fileobj(
+            Bucket=self.s3_bucket, Key=self.s3_key, Fileobj=target_file, ExtraArgs=extra_args
+        )
+        target_file.flush()
 
     def generate_presigned_url(self) -> str:
+        """
+        Generates a presigned url pointing to the code archive
+        """
         s3_client: "S3Client" = aws_stack.connect_to_service("s3", region_name="us-east-1")
         params = {"Bucket": self.s3_bucket, "Key": self.s3_key}
         if self.s3_object_version:
             params["VersionId"] = self.s3_object_version
         return s3_client.generate_presigned_url("get_object", Params=params)
 
+    def get_unzipped_code_location(self) -> Path:
+        """
+        Get the location of the unzipped archive on disk
+        """
+        return Path(f"{tempfile.gettempdir()}/lambda/{self.s3_bucket}/{self.id}/code")
```

Comment:
Since this is getting more complex now, what do you think about adding a short doc to the S3Code class summarizing the location + lifecycle of all the resources it creates/destroys? (like the path here)